WordPress/gutenberg
# Widgets editor: Improve compatibility with WP 5.8

## Description

Improves upon https://github.com/WordPress/gutenberg/pull/32176 and fixes https://github.com/WordPress/gutenberg/issues/32281.

Contains a few changes intended to make the Gutenberg plugin compatible with both versions of WordPress which support widgets (5.8) and versions that don't (5.7).

- Only load the PHP code responsible for implementing the widgets editor if core does not already have it.
- Do not link the Appearance → Widgets menu item to `wp-admin/theme.php?page=gutenberg-widgets` if core uses the widgets editor in `wp-admin/widgets.php`.
- Fix https://github.com/WordPress/gutenberg/issues/32281 by ensuring that we intercept the Legacy Widget preview iframe's request immediately after the block is registered.
- Remove the PHP unit tests for the widgets REST API endpoints. These tests exist in core now and will always fail when run in our CI environment which uses latest core `trunk`.

## How has this been tested?

With an environment that has both `wordpress-develop-git` and `gutenberg` cloned and running together:

1. Check out the `5.7` branch in `wordpress-develop-git`.
1. Disable the gutenberg plugin. 
1. Go to Appearance → Widgets and Appearance → Customize → Widgets. The old editors should appear and work.
1. Enable the gutenberg plugin.
1. Go to Appearance → Widgets and Appearance → Customize → Widgets. The new editors should appear and work. The URL should be `theme.php?page=gutenberg-widgets`.
1. Check out the `master` branch in `wordpress-develop-git`.
1. Disable the gutenberg plugin. 
1. Go to Appearance → Widgets and Appearance → Customize → Widgets. The new editors should appear and work. The URL should be `widgets.php`.
1. Enable the gutenberg plugin.
1. Go to Appearance → Widgets and Appearance → Customize → Widgets. The new editors should appear and work. The URL should be `widgets.php`.

Note: There is a PHP warning caused by registering `core/legacy-widget` twice when running the Gutenberg plugin against latest core `trunk`. This will be fixed once `@wordpress/block-library` is published to npm with the changes in this PR and included in core. You can verify this by copying `packages/block-library/src/legacy-widget/index.php` in Gutenberg to `blocks/legacy-widget.php` in core.
